File: .migration-backup/agent_pay_for_urself/adapters/market_data.py
"""Market data provider boundary.

The current implementation keeps a deterministic stub provider for tests and
offline runs, while allowing opt-in live market data through Yahoo Finance.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Callable, Mapping
from math import isnan
from typing import Any

from agent_pay_for_urself.schemas import MarketData

DEFAULT_PRICE = 100.0
DEFAULT_PE_RATIO = 20.0
DEFAULT_YAHOO_NEWS_COUNT = 3
DEFAULT_YAHOO_HISTORY_PERIOD = "5d"
YAHOO_FINANCE_SYMBOL_OVERRIDES = {
    "005930": "005930.KS",
}

YAHOO_TO_KIS_ORDER_EXCHANGE_CODE = {
    "nasdaq": "NASD",
    "nasdaqcm": "NASD",
    "nasdaqgm": "NASD",
    "nasdaqgs": "NASD",
    "nms": "NASD",
    "nas": "NASD",
    "nyse": "NYSE",
    "newyorkstockexchange": "NYSE",
    "nyq": "NYSE",
    "nys": "NYSE",
    "nyseamerican": "AMEX",
    "nysemkt": "AMEX",
    "amex": "AMEX",
    "americanstockexchange": "AMEX",
    "ase": "AMEX",
    "ams": "AMEX",
}


def normalize_yahoo_finance_symbol(symbol: str) -> str:
    """Return the Yahoo Finance lookup symbol for one normalized stock code.

    The public API keeps the original symbol for workflow outputs, but Yahoo
    Finance needs the exchange suffix for domestic Korea listings.
    """

    normalized_symbol = symbol.strip().upper()
    if not normalized_symbol:
        return normalized_symbol
    if normalized_symbol.endswith((".KS", ".KQ")):
        return normalized_symbol
    return YAHOO_FINANCE_SYMBOL_OVERRIDES.get(normalized_symbol, normalized_symbol)


class MarketDataProvider(ABC):
    """Provides normalized market data to the data collection agent."""

    mode_name = "configured"

    @abstractmethod
    def get_market_data(self, symbol: str) -> MarketData:
        """Return current market data for one already-normalized symbol."""


class StubMarketDataProvider(MarketDataProvider):
    """Deterministic provider used until real data clients are introduced."""

    mode_name = "stub"

    def get_market_data(self, symbol: str) -> MarketData:
        return MarketData(
            symbol=symbol,
            latest_price=DEFAULT_PRICE,
            broker_exchange_code=None,
            news_headlines=(f"{symbol} market update",),
            financial_metrics={"pe_ratio": DEFAULT_PE_RATIO},
        )


class YahooFinanceMarketDataProvider(MarketDataProvider):
    """Live market data provider backed by Yahoo Finance via ``yfinance``."""

    mode_name = "yahoo"

    def __init__(
        self,
        ticker_factory: Callable[[str], object] | None = None,
        news_count: int = DEFAULT_YAHOO_NEWS_COUNT,
        history_period: str = DEFAULT_YAHOO_HISTORY_PERIOD,
    ) -> None:
        self._ticker_factory = ticker_factory or self._build_default_ticker_factory()
        self._news_count = news_count
        self._history_period = history_period

    def get_market_data(self, symbol: str) -> MarketData:
        """Fetch and normalize the subset of Yahoo data used by the workflow."""

        ticker = self._ticker_factory(symbol)
        latest_price = self._extract_latest_price(ticker, symbol)
        news_headlines = self._extract_news_headlines(ticker)
        financial_metrics = self._extract_financial_metrics(ticker)
        broker_exchange_code = self._extract_broker_exchange_code(ticker)
        return MarketData(
            symbol=symbol,
            latest_price=latest_price,
            broker_exchange_code=broker_exchange_code,
            news_headlines=news_headlines,
            financial_metrics=financial_metrics,
        )

    def _build_default_ticker_factory(self) -> Callable[[str], object]:
        try:
            import yfinance as yf
        except ImportError as exc:
            raise RuntimeError(
                "Yahoo Finance provider requires the 'yfinance' package to be installed."
            ) from exc
        return yf.Ticker

    def _extract_latest_price(self, ticker: object, _symbol: str) -> float:
        fast_info = getattr(ticker, "fast_info", {}) or {}
        latest_price = self._coerce_optional_float(fast_info.get("lastPrice"))
        if latest_price is not None:
            return latest_price

        try:
            history = ticker.history(period=self._history_period)
        except Exception:
            return DEFAULT_PRICE
        if getattr(history, "empty", True):
            return DEFAULT_PRICE

        close_prices = history["Close"].dropna()
        if len(close_prices) == 0:
            return DEFAULT_PRICE
        return float(close_prices.iloc[-1])

    def _extract_news_headlines(self, ticker: object) -> tuple[str, ...]:
        news_items: list[Any]
        get_news = getattr(ticker, "get_news", None)
        try:
            if callable(get_news):
                news_items = list(get_news(count=self._news_count))
            else:
                news_items = list(getattr(ticker, "news", []) or [])
        except Exception:
            return ()

        headlines: list[str] = []
        for item in news_items:
            title = self._extract_news_title(item)
            if title is not None:
                headlines.append(title)
            if len(headlines) >= self._news_count:
                break
        return tuple(headlines)

    def _extract_financial_metrics(self, ticker: object) -> dict[str, float]:
        info = self._safe_ticker_info(ticker)
        if not isinstance(info, Mapping):
            return {}

        pe_ratio = self._coerce_optional_float(info.get("trailingPE"))
        if pe_ratio is None:
            pe_ratio = self._coerce_optional_float(info.get("forwardPE"))
        if pe_ratio is None:
            return {}
        return {"pe_ratio": pe_ratio}

    def _extract_broker_exchange_code(self, ticker: object) -> str | None:
        info = self._safe_ticker_info(ticker)
        if not isinstance(info, Mapping):
            return None

        for raw_value in (info.get("exchange"), info.get("fullExchangeName")):
            normalized = self._normalize_exchange_name(raw_value)
            if normalized is None:
                continue
            broker_exchange_code = YAHOO_TO_KIS_ORDER_EXCHANGE_CODE.get(normalized)
            if broker_exchange_code is not None:
                return broker_exchange_code
        return None

    def _extract_news_title(self, item: Any) -> str | None:
        if not isinstance(item, Mapping):
            return None

        direct_title = item.get("title")
        if isinstance(direct_title, str) and direct_title.strip():
            return direct_title.strip()

        content = item.get("content")
        if isinstance(content, Mapping):
            nested_title = content.get("title")
            if isinstance(nested_title, str) and nested_title.strip():
                return nested_title.strip()
        return None

    def _coerce_optional_float(self, value: Any) -> float | None:
        if value is None or isinstance(value, bool):
            return None
        if isinstance(value, (int, float)):
            numeric_value = float(value)
            if isnan(numeric_value):
                return None
            return numeric_value
        return None

    def _normalize_exchange_name(self, value: Any) -> str | None:
        if not isinstance(value, str):
            return None
        collapsed = "".join(character for character in value.lower() if character.isalnum())
        return collapsed or None

    def _safe_ticker_info(self, ticker: object) -> Mapping[str, Any]:
        """Return Yahoo ticker metadata when available and otherwise fall back to empty data.

        yfinance can raise for some symbols when Yahoo omits exchange timezone metadata or
        returns a partial quoteSummary payload. The workflow should keep going when price data
        is still available, so metadata lookups stay best effort instead of failing the request.
        """

        try:
            info = getattr(ticker, "info", {}) or {}
        except Exception:
            return {}
        if not isinstance(info, Mapping):
            return {}
        return info

File: artifacts/api-server/src/engine/market-data.ts
import { getMarketDataProviderName } from "../lib/runtime-config";
import type { MarketData } from "./schemas";

const DEFAULT_PRICE = 100.0;
const DEFAULT_PE_RATIO = 20.0;
const DEFAULT_YAHOO_NEWS_COUNT = 3;
const DEFAULT_YAHOO_HISTORY_PERIOD = "5d";
const DEFAULT_TIMEOUT_MS = 5000;

const STUB_PRICES: Record<string, number> = {
  AAPL: 189.5,
  MSFT: 415.2,
  TSLA: 242.8,
  NVDA: 875.4,
  AMZN: 185.1,
  GOOGL: 175.3,
  META: 498.7,
  NFLX: 635.2,
  "005930": 71200,
  "035420": 198000,
};

const STUB_PE: Record<string, number> = {
  AAPL: 28.5,
  MSFT: 34.2,
  TSLA: 65.1,
  NVDA: 42.8,
  AMZN: 38.4,
  GOOGL: 22.7,
  META: 25.3,
  NFLX: 45.6,
  "005930": 12.4,
  "035420": 32.1,
};

const STUB_NEWS: Record<string, string[]> = {
  AAPL: ["애플, 신형 AI 기능 탑재 아이폰 출시 예정", "애플 실적 발표 예상치 상회"],
  MSFT: ["마이크로소프트, Azure AI 수요 급증", "코파일럿 기업 도입 확대 보고"],
  TSLA: ["테슬라, 2분기 인도량 기대 이하", "테슬라 오토파일럿 규제 불확실성"],
  NVDA: ["엔비디아 데이터센터 GPU 수요 역대 최고", "블랙웰 아키텍처 출하 호조"],
  AMZN: ["아마존 AWS 클라우드 점유율 유지", "아마존 광고 매출 견조한 성장세"],
  GOOGL: ["구글 검색 AI 통합 가속화", "유튜브 광고 수익 개선"],
  META: ["메타 AI 어시스턴트 이용자 급증", "광고 단가 회복세 지속"],
};

const YAHOO_FINANCE_SYMBOL_OVERRIDES: Record<string, string> = {
  "005930": "005930.KS",
  "035420": "035420.KS",
};

const YAHOO_TO_KIS_ORDER_EXCHANGE_CODE: Record<string, string> = {
  nasdaq: "NASD",
  nasdaqcm: "NASD",
  nasdaqgm: "NASD",
  nasdaqgs: "NASD",
  nms: "NASD",
  nas: "NASD",
  nyse: "NYSE",
  newyorkstockexchange: "NYSE",
  nyq: "NYSE",
  nys: "NYSE",
  nyseamerican: "AMEX",
  nysemkt: "AMEX",
  amex: "AMEX",
  americanstockexchange: "AMEX",
  ase: "AMEX",
  ams: "AMEX",
  kospi: "005930",
  kosdaq: "035420",
};

export function normalizeYahooFinanceSymbol(symbol: string): string {
  const normalized = symbol.trim().toUpperCase();
  if (!normalized) return normalized;
  if (normalized.endsWith(".KS") || normalized.endsWith(".KQ")) return normalized;
  return YAHOO_FINANCE_SYMBOL_OVERRIDES[normalized] ?? normalized;
}

function normalizeExchangeName(value: unknown): string | null {
  if (typeof value !== "string") return null;
  const collapsed = value.toLowerCase().replace(/[^a-z0-9]/g, "");
  return collapsed || null;
}

function coerceNumber(value: unknown): number | null {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string" && value.trim()) {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return null;
}

function extractNewsTitle(item: unknown): string | null {
  if (!item || typeof item !== "object") return null;
  const record = item as Record<string, unknown>;
  const title = record.title;
  if (typeof title === "string" && title.trim()) return title.trim();
  const content = record.content;
  if (content && typeof content === "object") {
    const nested = (content as Record<string, unknown>).title;
    if (typeof nested === "string" && nested.trim()) return nested.trim();
  }
  return null;
}

async function fetchJson<T>(url: string): Promise<T | null> {
  try {
    const response = await fetch(url, {
      signal: AbortSignal.timeout(DEFAULT_TIMEOUT_MS),
      headers: {
        "User-Agent": "Mozilla/5.0",
        Accept: "application/json",
      },
    });
    if (!response.ok) {
      return null;
    }
    return (await response.json()) as T;
  } catch {
    return null;
  }
}

class YahooFinanceMarketDataProvider {
  readonly mode_name = "yahoo";

  async getMarketData(symbol: string): Promise<MarketData> {
    const normalizedSymbol = normalizeYahooFinanceSymbol(symbol);
    const [quotePayload, newsPayload] = await Promise.all([
      this.fetchQuote(normalizedSymbol),
      this.fetchNews(normalizedSymbol),
    ]);

    const quote = quotePayload?.quoteResponse?.result?.[0] ?? null;
    const latestPrice =
      coerceNumber(quote?.regularMarketPrice) ??
      coerceNumber(quote?.postMarketPrice) ??
      coerceNumber(quote?.preMarketPrice) ??
      DEFAULT_PRICE;

    const peRatio =
      coerceNumber(quote?.trailingPE) ?? coerceNumber(quote?.forwardPE) ?? DEFAULT_PE_RATIO;

    const exchangeSource = quote?.fullExchangeName ?? quote?.exchange ?? quote?.marketState;
    const normalizedExchange = normalizeExchangeName(exchangeSource);
    const brokerExchangeCode =
      normalizedExchange !== null ? YAHOO_TO_KIS_ORDER_EXCHANGE_CODE[normalizedExchange] ?? null : null;

    const newsHeadlines = this.extractHeadlines(newsPayload);

    return {
      symbol: symbol.trim().toUpperCase(),
      latest_price: latestPrice,
      broker_exchange_code: brokerExchangeCode,
      news_headlines: newsHeadlines,
      financial_metrics: { pe_ratio: peRatio },
    };
  }

  private async fetchQuote(symbol: string): Promise<YahooQuoteResponse | null> {
    const url = `https://query1.finance.yahoo.com/v7/finance/quote?symbols=${encodeURIComponent(symbol)}`;
    return fetchJson<YahooQuoteResponse>(url);
  }

  private async fetchNews(symbol: string): Promise<YahooNewsResponse | null> {
    const url = `https://query1.finance.yahoo.com/v1/finance/search?q=${encodeURIComponent(symbol)}&quotesCount=0&newsCount=${DEFAULT_YAHOO_NEWS_COUNT}`;
    return fetchJson<YahooNewsResponse>(url);
  }

  private extractHeadlines(payload: YahooNewsResponse | null): string[] {
    const newsItems = payload?.news ?? [];
    const headlines: string[] = [];
    for (const item of newsItems) {
      const title = extractNewsTitle(item);
      if (title) {
        headlines.push(title);
      }
      if (headlines.length >= DEFAULT_YAHOO_NEWS_COUNT) {
        break;
      }
    }
    return headlines;
  }
}

class StubMarketDataProvider {
  readonly mode_name = "stub";

  getMarketData(symbol: string): MarketData {
    const upper = symbol.toUpperCase();
    return {
      symbol: upper,
      latest_price: STUB_PRICES[upper] ?? DEFAULT_PRICE,
      broker_exchange_code: null,
      news_headlines: STUB_NEWS[upper] ?? [`${upper} market update`],
      financial_metrics: { pe_ratio: STUB_PE[upper] ?? DEFAULT_PE_RATIO },
    };
  }
}

type YahooQuoteResponse = {
  quoteResponse?: {
    result?: Array<{
      regularMarketPrice?: number | string;
      postMarketPrice?: number | string;
      preMarketPrice?: number | string;
      trailingPE?: number | string;
      forwardPE?: number | string;
      exchange?: string;
      fullExchangeName?: string;
      marketState?: string;
    }>;
  };
};

type YahooNewsResponse = {
  news?: Array<Record<string, unknown>>;
};

export type MarketDataProvider = StubMarketDataProvider | YahooFinanceMarketDataProvider;

export function buildMarketDataProvider(): MarketDataProvider {
  const providerName = getMarketDataProviderName();
  if (providerName === "yahoo" || providerName === "yfinance") {
    return new YahooFinanceMarketDataProvider();
  }
  return new StubMarketDataProvider();
}

export { StubMarketDataProvider, YahooFinanceMarketDataProvider };
